backend/src/main/java/com/vax/warden/security/LoginAttemptService.java

```diff
@@ -0,0 +1,60 @@
+package com.vax.warden.security;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import javax.servlet.http.HttpServletRequest;
+import org.springframework.stereotype.Service;
+
+@Service
+public class LoginAttemptService {
+    private final int MAX_ATTEMPTS = 3;
+    private final int BLOCKING_TIME_MINUTES = 20;
+    private LoadingCache<String, Integer> attemptsCache;
+
+    public LoginAttemptService() {
+        super();
+        attemptsCache =
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(BLOCKING_TIME_MINUTES, TimeUnit.MINUTES)
+                        .build(
+                                new CacheLoader<String, Integer>() {
+                                    public Integer load(String key) {
+                                        return 0;
+                                    }
+                                });
+    }
+
+    public static String getClientIP(HttpServletRequest request) {
+        final String xfHeader = request.getHeader("X-Forwarded-For");
+        if (xfHeader != null) {
+            return xfHeader.split(",")[0];
+        }
+        return request.getRemoteAddr();
+    }
+
+    public void loginSucceeded(String key) {
+        attemptsCache.invalidate(key);
+    }
+
+    public void loginFailed(String key) {
+        int attempts = 0;
+        try {
+            attempts = attemptsCache.get(key);
+        } catch (ExecutionException e) {
+            attempts = 0;
+        }
+        attempts++;
+        attemptsCache.put(key, attempts);
+    }
+
+    public boolean isBlocked(String key) {
+        try {
+            return attemptsCache.get(key) >= MAX_ATTEMPTS;
```

Comment:
should this be `>` instead of `>=` since it's maximum allowed attempts I presume?

Reply:
It works with `>=` since once `MAX_ATTEMPTS` is reached, the IP should be blocked. E.g. 
- failed attempts = 1: don't block next attempt
- failed attempts = 2: don't block next attempt
- failed attempts = 3: BLOCK *next* attempt
Hence the 4th attempt is blocked, despite being potentially successful.

"If an IP address performs 3 consecutive failed authentication attempts, block it for a given amount of time (e.g., 20 mins)."

Reply:
alright thanks for clarifying, this makes sense